adr/0009-native-class-initialization.adoc

```diff
@@ -0,0 +1,330 @@
+
+== Context and Problem Statement
+
+Build time initialization of classes in native mode aligns with Quarkus' philosophy of doing as much as possible at build time to optimize run time performance and reduce startup time.
+However, build time initialization often leads to issues that require additional configuration or even code substitutions to be worked around.
+This complicates the development process and discourages developers from using Quarkus' native mode due to increased complexity if they rely on libraries not handled by Quarkus or some of its extensions.
+
+Quarkus applications may depend at run time on:
+
+* Quarkus core and core extensions, which are distributed under the `io.quarkus` group ID.
+* Quarkiverse extensions, which may be distributed under the `io.quarkiverse` or another group ID.
+* User provided extensions not listed above
+* Third-party libraries
+
+The goal of this ADR is to define:
+1. which of the above should be initialized at build time and which should be initialized at run time by default in native mode.
+2. how developers can opt in or out of build time initialization.
+
+== Decision Drivers
+
+* Reduce entry barrier for developers that want to use Quarkus native mode
+* Benefit from build time initialization when possible
+* Allow developers to opt-in to build time initialization of parts that are run time initialized by default
+
+== Considered options for default class initialization policy
+
+[source, mermaid]
+....
+---
+title: Legend
+config:
+  look: handDrawn
+  theme: default
+---
+graph TD
+  BT[Build-time initialized]
+  RT[Run-time initialized]
+
+classDef buildtime fill:#bbf;
+class B,C,BT buildtime
+....
+
+=== Initialize all classes at build time
+
+In this scenario, all classes in the Quarkus application are build time initialized, unless explicitly configured otherwise.
+This option has been the default behavior in Quarkus for a long time.
+
+[source, mermaid]
+....
+---
+config:
+  look: handDrawn
+  theme: default
+---
+graph LR
+  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
+  QA --> arc[io.quarkus.arc:arc] --> R1[...]
+  arc --> jboss[org.jboss.logging:jboss-logging]
+  QA --> core[io.quarkus:quarkus-core] --> R2[...]
+  A --> UserExtension[com.example:my-extension] --> R3[...]
+  UserExtension --> core
+  UserExtension --> jboss
+  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
+  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
+  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]
+
+classDef white fill:#fff;
+class A white
+classDef buildtime fill:#bbf;
+class QA,arc,core,Camel,R1,R2,R3,R4,R5,UserExtension,UserExtensionBT,RestAssured,jboss buildtime
+....
+
+
+*Pros*:
+
+- It allows Quarkus to take advantage of build time initialization as much as possible.
+
+*Cons*:
+
+- It requires developers to deal with issues that arise from build time initialization that are often non-trivial.
```

Comment:
When we talk about "developers", could we spell out if we're referring to our users vs the Quarkus code maintainers? The extension developers?  Sometimes the context makes it clear, but for example in this case I'm not sure what is being stated.